src/quic/tokens.cc

```diff
@@ -0,0 +1,255 @@
+
+StatelessResetToken::StatelessResetToken(const TokenSecret& secret,
+                                         const CID& cid)
+    : ptr_(buf_) {
+  CHECK_EQ(ngtcp2_crypto_generate_stateless_reset_token(
+               buf_, secret, kStatelessTokenLen, cid),
+           0);
+}
+
+StatelessResetToken::StatelessResetToken(uint8_t* token,
+                                         const TokenSecret& secret,
+                                         const CID& cid)
+    : ptr_(token) {
+  CHECK_EQ(ngtcp2_crypto_generate_stateless_reset_token(
+               token, secret, kStatelessTokenLen, cid),
+           0);
+}
+
+StatelessResetToken::StatelessResetToken(const StatelessResetToken& other)
+    : ptr_(buf_) {
+  if (other) {
+    memcpy(buf_, other.ptr_, kStatelessTokenLen);
+  } else {
+    ptr_ = nullptr;
+  }
+}
+
+StatelessResetToken::operator const uint8_t*() const {
+  return ptr_ != nullptr ? ptr_ : buf_;
+}
+
+StatelessResetToken::operator const char*() const {
+  return reinterpret_cast<const char*>(ptr_ != nullptr ? ptr_ : buf_);
+}
+
+StatelessResetToken::operator bool() const {
+  return ptr_ != nullptr;
+}
+
+bool StatelessResetToken::operator==(const StatelessResetToken& other) const {
+  if (ptr_ == other.ptr_) return true;
+  if ((ptr_ == nullptr && other.ptr_ != nullptr) ||
+      (ptr_ != nullptr && other.ptr_ == nullptr)) {
+    return false;
+  }
+  return memcmp(ptr_, other.ptr_, kStatelessTokenLen) == 0;
+}
+
+bool StatelessResetToken::operator!=(const StatelessResetToken& other) const {
+  return !(*this == other);
+}
+
+std::string StatelessResetToken::ToString() const {
+  if (ptr_ == nullptr) return std::string();
+  char dest[kStatelessTokenLen * 2];
+  size_t written =
+      StringBytes::hex_encode(*this, kStatelessTokenLen, dest, arraysize(dest));
+  DCHECK_EQ(written, arraysize(dest));
+  return std::string(dest, written);
+}
+
+size_t StatelessResetToken::Hash::operator()(
+    const StatelessResetToken& token) const {
+  size_t hash = 0;
+  if (token.ptr_ == nullptr) return hash;
+  for (size_t n = 0; n < kStatelessTokenLen; n++)
+    hash ^= std::hash<uint8_t>{}(token.ptr_[n]) + 0x9e3779b9 + (hash << 6) +
+            (hash >> 2);
+  return hash;
+}
+
+StatelessResetToken StatelessResetToken::kInvalid;
+
+// ============================================================================
+// RetryToken and RegularToken
+namespace {
+ngtcp2_vec generate_retry_token(uint8_t* buffer,
+                                uint64_t version,
+                                const SocketAddress& address,
+                                const CID& retry_cid,
+                                const CID& odcid,
+                                const TokenSecret& token_secret) {
+  ssize_t ret =
+      ngtcp2_crypto_generate_retry_token(buffer,
+                                         token_secret,
+                                         TokenSecret::QUIC_TOKENSECRET_LEN,
+                                         version,
```

Comment:
According to the [documentation of `ngtcp2_crypto_generate_retry_token`](https://nghttp2.org/ngtcp2/ngtcp2_crypto_generate_retry_token.html), the fourth argument `version` is of type `uint32_t`. Is that incorrect, is this an intended narrowing conversion, or should the type of `version` be changed from `uint64_t` to `uint32_t`?